utils/indicators.py

```diff
@@ -0,0 +1,304 @@
+
+    def _load_indices(self, indices_file):
+        # Read index configurations
+        with open(indices_file, "r") as stream:
+            try:
+                return yaml.safe_load(stream)
+            except yaml.YAMLError as e:
+                logging.error(e)
+                return None
+
+    def generate_index(self, index_config, year):
+        """
+        Generates an index based on the provided index configuration.
+
+        Args:
+            index_config (dict): Configuration for generating the index.
+
+        Returns:
+            ee.Image: The generated index clipped to the region of interest.
+        """
+
+        # Calculate date range, assume 1 year
+        start_date = str(datetime.date(year, 1, 1))
+        end_date = str(datetime.date(year, 12, 31))
+        daterange = [start_date, end_date]
+
+        # Calculate index based on type
+        match index_config["gee_type"]:
+            case "image":
+                dataset = ee.Image(index_config["gee_path"]).clip(self.roi)
+                if index_config.get("select"):
+                    dataset = dataset.select(index_config["select"])
+            case "image_collection":
+                dataset = (
+                    ee.ImageCollection(index_config["gee_path"])
+                    .filterBounds(self.roi)
+                    .map(lambda image: image.clip(self.roi))
+                    .mean()
+                )
+                if index_config.get("select"):
+                    dataset = dataset.select(index_config["select"])
+            case "feature_collection":
+                dataset = (
+                    ee.Image()
+                    .float()
+                    .paint(
+                        ee.FeatureCollection(index_config["gee_path"]),
+                        index_config["select"],
+                    )
+                    .clip(self.roi)
+                )
+            case "algebraic":
+                image = self._cloudfree(index_config["gee_path"], daterange)
+                # to-do: params should come from index_config
+                dataset = image.normalizedDifference(["B4", "B3"])
+            case _:
+                dataset = None
+
+        if not dataset:
+            raise Exception("Failed to generate dataset.")
+
+        # Whether to display on GEE map
+        if self.show and index_config.get("show"):
+            map.addLayer(dataset, index_config["viz"], index_config["name"])
+
+        logging.info(f"Generated index: {index_config['name']}")
+        return dataset
+
+    def zonal_mean_index(self, index_key, year):
+        index_config = self.indices[index_key]
+        dataset = self.generate_index(index_config, year)
+        # zm = self._zonal_mean(single, index_config.get('bandname') or 'constant')
+        out = dataset.reduceRegion(
+            **{
+                "reducer": ee.Reducer.mean(),
+                "geometry": self.roi,
+                "scale": 200,  # map scale
+            }
```

Comment:
It seems the high latency comes from this function for the "Air" index.

https://developers.google.com/earth-engine/guides/scale

Reply:
Wasn't able to resolve latency issue, so I removed "Air" from the index list for now.